### ISSUE TITLE
WASAPI: enable automatic samplerate conversion

### ISSUE THREAD

**Issue by Spacechild1** [supercollider/supercollider#6506]
## Motivation

Currently, if you want to use the WASAPI audio API, you have to pass the system samplerate, otherwise the Server refuses to boot. For example:
```
Requested sample rate 44100.000000 for devices Windows WASAPI : IN 1-2 (2- BEHRINGER UMC 404HD 192k) and Windows WASAPI : OUT 1-2 (2- BEHRINGER UMC 404HD 192k) is not supported.
SC_PortAudioDriver: PortAudio failed at Pa_OpenStream with error: 'Invalid sample rate'
```

## Description of Proposed Feature

Actually, WASAPI has an option for automatic samplerate conversion, which is also exposed by portaudio: https://github.com/PortAudio/portaudio/blob/57aa393109ec996799d3a5846c9ecb0a65b64644/include/pa_win_wasapi.h#L81-L85

## Plan for Implementation

I have already enabled this feature in Pd's portaudio backend and we could the same for scsynth and supernova:

https://github.com/pure-data/pure-data/blob/5305f2f30aed412bd77f36468e4318cb65391509/src/s_audio_pa.c#L243-L264

**Comment by dyfer** [comment]
Thanks for reporting this. I remember encountering something like this when working on the PortAudio upgrade in SC a while back.

For reference, with coreaudio and jack backends, passing no samplerate will cause the server to boot to the samplerate the hardware is running at. I feel like this is the desired default behavior. Unfortunately, PA/asio does not behave the same functionality and in fact PA author(s) have a different understanding what a "default" sample rate is. In any case, I would personally prefer for the default behavior to be that the backend gets the current samplerate from the audio api and boots to that. Would there be a way to do it that way with PA/WASAPI?

Also, for SC as a "serious" audio app, I feel a bit apprehensive of introducing a "hidden" samplerate conversion. At the same time, for amateur use it might not matter as much...

Finally, not directly related, but I was wondering if we could eventually disable DirectX and MME backends on Windows (maybe WDM-KS too?) to make the audio hardware selection easier. Of course an issue like this one would need to be addressed first.

**Comment by Spacechild1** [comment]
> For reference, with coreaudio and jack backends, passing no samplerate will cause the server to boot to the samplerate the hardware is running at.

That's also how it works with WASAPI on portaudio! For some reason, ASIO defaults to 44100, even if the hardware is currently running at 48000. (This is even mentioned in the docs, so it's a known issue.) I need to check if this is a mistake on our end or if it's a bug in portaudio (which might have already been fixed in recent releases).

---

However, this issue is adressing a slightly different problem:

Some projects require/assume a specific samplerate. With ASIO you can simply request a sample rate and the driver will typically set it for you (as long as no other app is currently running under a different sample rate). However, for WASAPI you need to go to the (arcane) Windows audio settings... In this case it would be great if it would simply resample to the desired rate. 

> EDIT: I'm guessing MME, which is chosen by default, does SR conversion anyway... so maybe avoiding SR conversion is not a valid argument for a "consumer" API. If we want not-SR-converted output, we'd use ASIO anyway...

This exactly what I had in mind when I added WASAPI support for Pd! In fact, we deprecated the old MME backend and made WASAPI the default, so we really needed to have the same automatic SR conversion behavior. As you said, on Windows you would/should always use ASIO in a professional setting.

> Finally, not directly related, but I was wondering if we could eventually disable DirectX and MME backends on Windows

Yeah, I think there is no real reason to use any of these in 2024... WASAPI and ASIO (professional) is really all you need.

**Comment by dyfer** [comment]
> For some strange reason, ASIO defaults to 44100, even if the hardware is currently running at 48000. (This is even mentioned in the docs, so it's a known issue.) I need to check if this is a mistake on our end or if it's a bug in portaudio (which might have already been fixed upstream.)

This is PA's convention, I had some exchanges about this with Ross Bencina on the mailing list. Unfortunately full archive is not available online, part of his reply to getting the samplerate from what the device currently runs at was:

> ASIOGetSampleRate returns whatever the current ASIO device sample rate
is set to. This really could be be anything -- whatever state the last
program left the device in, or whatever setting was selected on the
front panel (for some devices). Neither of these satisfies any
definition of what PortAudio would consider a default sample rate --
which should be some immutable property of the device.

That's not my understanding of what the default SR should be.

In any case, [what PA does is that it goes through the hardcoded list of samplerates](https://github.com/PortAudio/portaudio/blob/master/src/hostapi/asio/pa_asio.cpp#L1041-L1052) - `{44100.0, 48000.0, 32000.0, 24000.0, 22050.0, 88200.0, 96000.0, 192000.0, 16000.0, 12000.0, 11025.0, 9600.0, 8000.0 }` - asking the device whether it _supports_ the given sample rate. So in 99.9% of the cases we'd get a git on the first one on that list - 44100. The only time this would not be the case is if the device was "locked" to another SR.

I inquired about getting the samplerate from the device, and after a lot of back and forth the reply seemed to be that this won't be supported internally by PA, but a possible way forward is to add another function to `PaAsio_` that checks for the current SR of the device.

**Comment by dyfer** [comment]
> However, for WASAPI you need to go to the (arcane) Windows audio settings...

So with WASAPI we can't set the SR of the device? That's quite bizarre...
Is that the case with MME as well, do you know?

**Comment by Spacechild1** [comment]
> So with WASAPI we can't set the SR of the device? That's quite bizarre...

At least not in shared mode. Which makes sense. Maybe it can in exclusive mode (which is disabled by default).

> Is that the case with MME as well, do you know?

I think MME just uses the system sample rate and resamples as needed. There is no exlusive access to the audio hardware.

**Comment by Spacechild1** [comment]
> That's not my understanding of what the default SR should be.

Yeah, mine neither... In particular, I don't buy the following argument:

>  Neither of these satisfies any definition of what PortAudio would consider a default sample rate -- which should be some immutable property of the device.

Windows' system sample rate or Jack's sample rate isn't immutable either, yet PA uses it as the default sample rate.

In the case of ASIO we have the weird situation that the "default sample rate" might not even work! This happens when another client is already running with a different sample rate. When I run Pd @ 48 kHz and try to start SC with the "default sample rate" (44.1 kHz), the Server hangs...

> but a possible way forward is to add another function to PaAsio_ that returns the current SR of the ASIO device.

+1.

FWIW, the corresponding ASIO function is `ASIOGetSampleRate`. 

Personally, I would still prefer it if PA would just use the current device sample rate as the default sample rate.

**Comment by dyfer** [comment]
> Personally, I would still prefer it if PA used the current device sample rate as the default sample rate, as the current behavior just doesn't make any sense.

Me too. But Ross didn't seem to be amenable to that. 
Maybe it's worth it raising this again as a GH issue - the exchanges I had happened before the development moved to GH.

> > So with WASAPI we can't set the SR of the device? That's quite bizarre...
> 
> At least not in shared mode. Which makes sense.

Coreaudio allows setting the device sample rate, even though it works by default in a mode similar to WASAPI shared mode, IIUC. I find that reasonable and quite flexible.

But that's beyond point. In general I agree thet it would be desirable to activate the resampler. And then maybe disable other APIs in the Windows release builds later on (separate issue/PR)

**Comment by Spacechild1** [comment]
> Coreaudio allows setting the device sample rate, even though it works by default in a mode similar to WASAPI shared mode, IIUC. I find that reasonable and quite flexible.

Does it really set the device sample rate? How would that even work in shared mode? What happens to clients that are already running at the old sample rate? Or does this only work when there are no other clients? I'm curious!

**Comment by Spacechild1** [comment]
> Maybe it's worth it raising this again as a GH issue

I can do that. I might also make a PR for the `PaAsio_GetSampleRate()` function; it would be useful either way.

**Comment by dyfer** [comment]
> Does it really set the device sample rate? How would that even work in shared mode? What happens to clients that are already running at the old sample rate? Or does this only work when there are no other clients? I'm curious!

Yes, it does actually set the device sample rate, if desired.

When opening the stream, the client may request to resample, or to set the device to the designated SR (and probably set some other options as well). BTW we tell coreaudio to set the sample rate in PA on macOS in supernova using [this option](https://portaudio.com/docs/v19-doxydocs-dev/pa__mac__core_8h.html#abec61ce0294900be890928cd0907bcf6). In contrast, I think most "consumer" apps just request resampling.

When SR change happens, what happens with other streams is that they get temporarily broken (the clients might be notified, I don't really know the API), coreaudio enables resampling on existing streams and they resume. For example, if there's music playing through my soundcard and I start a DAW which sets my soundcard to a particular SR (or start SC with `s.options.sampleRate_()` set), then the music will momentarily silence and then resume playing, now using resampling, while the DAW finishes booting. At this point the soundcard runs at the new SR.

**Comment by Spacechild1** [comment]
Thanks, that's very interesting! Yes, that's definitely better than on Windows - as almost everything that is related to audio :)

**Comment by dyfer** [comment]
> When I run Pd @ 48 kHz and try to start SC with the "default sample rate" (44.1 kHz), the Server hangs...

I think this is a great example to bring up as an issue in PA.

**Comment by Spacechild1** [comment]
I think I came up with a better solution: https://github.com/PortAudio/portaudio/issues/970

Alternatively, here's a PR for `PaAsio_GetSampleRate`: See https://github.com/PortAudio/portaudio/pull/971

**Comment by Spacechild1** [comment]
Oh, thanks for the reminder, I totally forgot about this. I'm busy for the next 2 weeks, but after that I will try to make a PR. Feel free to ping me again if I forget :)

**Comment by Spacechild1** [comment]
Here you go: https://github.com/supercollider/supercollider/pull/6899
